### PR TITLE
Fix query on hasOne referenced model

### PR DESCRIPTION
**Pull request by georgehristov**
Cherrypick from #683

### CODE REVIEW

File: src/Persistence/Sql.php [hunk 1]
```diff
@@ -628,8 +632,16 @@ public function action(Model $model, $type, $args = [])
...
 
-                $fx = $args[0];
-                $field = is_string($args[1]) ? $model->getField($args[1]) : $args[1];
+                [$fx, $field] = $args;
+
+                if (is_string($field)) {
+                    if ($model->hasField($field)) {
+                        $field = $model->getField($field);
+                    } else {
+                        $field = new Expression($field);
```
Comment (mvorisek) [inline src/Persistence/Sql.php]: real usecases? we probably need to support args as well

File: tests/ReferenceTest.php [hunk 2]
```diff
@@ -71,6 +95,27 @@ public function testModelProperty()
...
+        $company->addField('name');
+        $user->hasOne('Company', [$company, 'our_field' => 'company_id', 'their_field' => 'id']);
+
+        $order = new Model($this->db, 'order');
+        $order->addField('company_id');
+        $order->addField('amount', ['default' => 20]);
+
+        $company->hasMany('Orders', [$order]);
+
+        $this->assertEquals(20, $user->ref('Company')->ref('Orders')->get('amount')); // 'amount' default value
```
Comment (DarkSide666) [inline tests/ReferenceTest.php]: Useless test - it will always return default value

What you should do I think is:
1. move this test to ReferenceSqlTest class and revert these changes from ReferenceTest
2. instead of get(amount) use count(...->export()) or export([amount]) or sum(amount) or something like that to actually test which records will be selected
3. Also add tests for unloaded and loaded models
```
$user->unload()->ref('Company')->ref('Orders')->export(['amount']);
$user->load(1)->ref('Company')->ref('Orders')->export(['amount']);
```

Comment (mvorisek) [inline tests/ReferenceTest.php]: where is the discussed usecase?

this test seems quite complex but your usecase was about one very simple thing

this test has only one assertion for a default val

File: src/Persistence/Sql.php [hunk 3]
```diff
@@ -620,6 +620,10 @@ public function action(Model $model, $type, $args = [])
                 $this->initQueryConditions($model, $query);
                 $this->setLimitOrder($model, $query);
 
+                if ($model->loaded()) {
+                    $query->where($model->id_field, $model->id);
```
Comment (mvorisek) [inline src/Persistence/Sql.php:624]: action materializes/render, so this makes sense (like `action('update')` uses the current ID)

but `action('select')` does not :( (so action should probably ignore loaded ID - how is/should `action()` behaviour be defined?)

before merge, confirm, that all uses of `action('field')` expect restricted field - does it make any sense to be used somewhere without the restriction to id?

File: src/Reference/HasOne.php [hunk 4]
```diff
@@ -225,10 +225,15 @@ public function ref($defaults = []): Model
...
+        $theirModel->reload_after_save = false;
+
         $theirModel->onHook(Model::HOOK_AFTER_SAVE, function ($theirModel) {
-            $ourValue = $this->their_field ? $theirModel->get($this->their_field) : $theirModel->id;
+            $theirValue = $this->their_field ? $theirModel->get($this->their_field) : $theirModel->id;
+
+            $this->getOurField()->set($theirValue)->owner->save();
 
-            $this->getOurModel()->set($this->our_field, $ourValue);
+            $theirModel->reload();
```
Comment (DarkSide666) [inline src/Reference/HasOne.php:238]: I think this is wrong.
Imagine we have `$user->hasOne(City::class)->addTitle()`, so it's `user.city_id = city.id` in SQL.

Now I change name of users city from Riga to London and save it.
What happens then? There is no need to save ourModel if theirField was not the one changed (dirty) in theirModel.

Why we need to reload city model (theirModel)? What maybe should be reloaded is users (ourModel) in case of some linked (title_field) or calculated fields.

Does it matter if ourModel is already loaded or not?

File: tests/ReferenceTest.php (hunk 2 above)
Comment (georgehristov) [reply tests/ReferenceTest.php]: tests expanded now from the #683

File: src/Reference/HasOneSql.php [hunk 5]
```diff
@@ -54,12 +54,15 @@ function (Model $ourModel) use ($theirFieldName) {
                     return $ourModel->refLink($this->link)->action('field', [$theirFieldName])->reset('order');
                 },
             ],
-            $defaults
+            $defaults,
```
Comment (DarkSide666) [inline src/Reference/HasOneSql.php:57]: Am I right  that there are no functional changes in HasOneSql class? Only some prettifying.

**Review by DarkSide666 — CHANGES_REQUESTED** [review]
See comments. Also not sure what exactly is fixed and where.

File: src/Persistence/Sql.php (hunk 1 above)
Comment (DarkSide666) [reply src/Persistence/Sql.php]: There definitely are usecases for Expression.

File: tests/ReferenceTest.php (hunk 2 above)
Comment (DarkSide666) [reply tests/ReferenceTest.php]: Good. Now also need to test same things with unloaded user model.
It should return all orders of all users.

File: src/Persistence/Sql.php (hunk 1 above)
Comment (mvorisek) [reply src/Persistence/Sql.php]: but it need args!

File: tests/ReferenceTest.php (hunk 2 above)
Comment (georgehristov) [reply tests/ReferenceTest.php]: Done

File: src/Persistence/Sql.php (hunk 1 above)
Comment (georgehristov) [reply src/Persistence/Sql.php]: Reverted to original functionality

Comment (mvorisek) [reply src/Persistence/Sql.php]: what is currently the best alternative /wo the planned optional trait?

File: tests/ReferenceSqlTest.php [hunk 6]
```diff
@@ -350,6 +350,56 @@ public function testOtherAggregates()
...
+        $user->load(1);
+
+        $this->assertSame([
+            ['id' => '1', 'company_id' => '1', 'description' => 'Vinny Company Order 1', 'amount' => '50'],
+            ['id' => '3', 'company_id' => '1', 'description' => 'Vinny Company Order 2', 'amount' => '15'],
+        ], $user->ref('Company')->ref('Orders')->export());
+
+        $user->unload();
+
+        $this->assertSame([
```
Comment (mvorisek) [inline tests/ReferenceSqlTest.php]: so @DarkSide666 was right with his guess yesterday? :)

File: tests/ReferenceSqlTest.php [hunk 7]
```diff
@@ -350,6 +350,56 @@ public function testOtherAggregates()
         $this->assertNull($l->get('chicken5'));
     }
 
+    public function testReferenceHasOneTraversing()
```
Comment (mvorisek) [inline tests/ReferenceSqlTest.php:353]: and hasMany?

File: src/Reference/HasOne.php (hunk 4 above)
Comment (georgehristov) [reply src/Reference/HasOne.php:238]: I made 2 changes here:
- delay the reloading of $theirModel to after setting $ourFieldValue to $theirFieldValue.
This is necesary as the `reload` cannot find the new city record (query on city is limited to 'only cities whose user.city_id = new city.id'. But there are no such users as the user.city_id has not been set to city.id yet. So we first set user.city_id=city.id and then reload city.
- save $ourModel when $ourFieldValue updated (original code was setting it but not saving). Now I added also check to only save when changed

File: tests/ReferenceSqlTest.php (hunk 7 above)
Comment (georgehristov) [reply tests/ReferenceSqlTest.php:353]: There are tests for hasMany already

File: tests/ReferenceSqlTest.php (hunk 6 above)
Comment (georgehristov) [reply tests/ReferenceSqlTest.php]: No, I was right. See the updated tests.

File: tests/ReferenceTest.php (hunk 2 above)
Comment (georgehristov) [reply tests/ReferenceTest.php]: All updated

File: tests/ReferenceSqlTest.php (hunk 6 above)
Comment (mvorisek) [reply tests/ReferenceSqlTest.php]: I see, is cloning on traversing documented?

File: src/Reference/HasOneSql.php (hunk 5 above)
Comment (georgehristov) [reply src/Reference/HasOneSql.php:57]: Only one major change in `HasOneSql::ref`. The below code is now executed irrespective if ourModel loaded or not (originally was only when not loaded)
```
$ourFieldExpression = $ourModel->action('field', [$ourField]);

return $theirModel->addCondition($theirField, $ourFieldExpression);
```

File: tests/ReferenceSqlTest.php (hunk 6 above)
Comment (mvorisek) [reply tests/ReferenceSqlTest.php]: documented here - https://github.com/atk4/data/blame/701532620fc597835002c2f7cfc58626e032bc26/docs/references.rst#L82

one central clone, that is good, but:
- the method should be named more like `getTheirModelClone()` or `createTheirModelClone`
- is it wanted that when model is created by Closure, the model can be not cloned?

Comment (georgehristov) [reply tests/ReferenceSqlTest.php]: The theirModel is generated every time you call `getTheirModel` so the name is correct.
The wording in the phpdoc is wrong. It is not a clone but a fresh object

Comment (mvorisek) [reply tests/ReferenceSqlTest.php]: yes, clone or factory... can be a clone, not a fresh object

Comment (georgehristov) [reply tests/ReferenceSqlTest.php]: Cannot be a clone as conditions are assigned to it depending on whether ourModel is loaded or not.

It is good as it is now. I have only solved the issue with referencing.

No need to waste time now on discussion abt method names, etc

File: src/Persistence/Sql.php (hunk 3 above)
Comment (DarkSide666) [reply src/Persistence/Sql.php:624]: Not 100% sure about this. Maybe there is some use-case where this is supposed to generate
`SELECT field FROM table` without ID condition even if model is already loaded.

File: src/Reference/HasOne.php (hunk 4 above)
Comment (DarkSide666) [reply src/Reference/HasOne.php:238]: Ok, probably that's correct.

**Review by DarkSide666 — APPROVED** [review]

File: src/Persistence/Sql.php (hunk 3 above)
Comment (mvorisek) [reply src/Persistence/Sql.php:624]: That is why I messages you about these changes - I think this is ok, as user can easily unload (or even clone + unload) and then call this method...

Comment (georgehristov) [reply src/Persistence/Sql.php:624]: No issues have been observed so far in testing, we will see in realtime use but this is true - model unload should be the solution if such usage exists
